File: skills/moltoffer-auto-apply/scripts/auto_apply.py
#!/usr/bin/env python3
"""
LinkedIn Easy Apply Automation Script

This script automates LinkedIn Easy Apply using Playwright.
Unknown questions are logged and the job is marked for human review.

Usage:
    python auto_apply.py [--yolo] [--limit N]

Options:
    --yolo      Apply without confirmation for each job
    --limit N   Maximum number of jobs to apply (default: unlimited)
"""

import argparse
import json
import re
import sys
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

import yaml
from playwright.sync_api import Page, sync_playwright, TimeoutError as PlaywrightTimeout

# Paths
SCRIPT_DIR = Path(__file__).parent
SKILL_DIR = SCRIPT_DIR.parent
DATA_DIR = SKILL_DIR / "data"
CANDIDATE_DIR = SKILL_DIR.parent / "moltoffer-candidate"

# Files
KNOWLEDGE_FILE = DATA_DIR / "knowledge.json"
APPLIED_FILE = DATA_DIR / "applied.json"
PENDING_FILE = DATA_DIR / "pending-jobs.json"
LOGS_FILE = DATA_DIR / "logs.json"
PERSONA_FILE = CANDIDATE_DIR / "persona.md"
CREDENTIALS_FILE = CANDIDATE_DIR / "credentials.local.json"


class KnowledgeBase:
    """Manages form answers and learns from AI responses."""

    def __init__(self):
        self.data = self._load()

    def _load(self) -> dict:
        if KNOWLEDGE_FILE.exists():
            return json.loads(KNOWLEDGE_FILE.read_text())
        return {"version": "1.0", "commonAnswers": {}, "assumptions": []}

    def save(self):
        KNOWLEDGE_FILE.write_text(json.dumps(self.data, indent=2, ensure_ascii=False))

    def get_answer(self, question: str) -> Optional[str]:
        """Look up answer for a question."""
        q_lower = question.lower()

        # Check common answers
        for key, value in self.data.get("commonAnswers", {}).items():
            if key.lower() in q_lower or q_lower in key.lower():
                return value

        # Check assumptions
        for assumption in self.data.get("assumptions", []):
            if assumption.get("question", "").lower() == q_lower:
                return assumption.get("assumedAnswer")

        return None

    def add_assumption(self, question: str, answer: str, reasoning: str):
        """Record a new assumed answer."""
        self.data.setdefault("assumptions", []).append({
            "question": question,
            "assumedAnswer": answer,
            "reasoning": reasoning,
            "timestamp": datetime.now(timezone.utc).isoformat()
        })
        self.save()
        print(f"  [LEARNED] Q: {question[:50]}... A: {answer}")


class PersonaInfo:
    """Extracts info from persona.md for form filling."""

    def __init__(self):
        self.raw = ""
        self.frontmatter = {}
        self._load()

    def _load(self):
        if not PERSONA_FILE.exists():
            raise FileNotFoundError(f"Persona file not found: {PERSONA_FILE}")

        content = PERSONA_FILE.read_text()
        self.raw = content

        # Parse YAML frontmatter
        if content.startswith("---"):
            parts = content.split("---", 2)
            if len(parts) >= 3:
                self.frontmatter = yaml.safe_load(parts[1]) or {}

    def get_info(self, key: str) -> Optional[str]:
        """Get info from persona."""
        # Check frontmatter first
        if key in self.frontmatter:
            return str(self.frontmatter[key])

        # Search in raw content
        patterns = {
            "name": r"###\s*Name\s*\n+([^\n]+)",
            "email": r"[\w.+-]+@[\w-]+\.[\w.-]+",
            "phone": r"\+?[\d\s()-]{10,}",
            "location": r"###\s*Current Location\s*\n+([^\n]+)",
        }

        if key in patterns:
            match = re.search(patterns[key], self.raw)
            if match:
                return match.group(1) if match.lastindex else match.group(0)

        return None


class UnknownQuestions:
    """Tracks unknown questions that need human answers."""

    def __init__(self):
        self.questions = []

    def add(self, question: str, options: list[str] = None):
        """Record an unknown question."""
        self.questions.append({
            "question": question,
            "options": options,
            "timestamp": datetime.now(timezone.utc).isoformat()
        })
        print(f"  [UNKNOWN] {question}")
        if options:
            print(f"    Options: {', '.join(options[:5])}")

    def has_unknowns(self) -> bool:
        return len(self.questions) > 0

    def print_summary(self):
        if not self.questions:
            return
        print(f"\n[ACTION REQUIRED] {len(self.questions)} unknown questions need answers:")
        print("Add these to knowledge.json, then re-run the script.\n")
        for q in self.questions:
            print(f"  Q: {q['question']}")
            if q.get('options'):
                print(f"     Options: {', '.join(q['options'][:5])}")


class ApplicationLogger:
    """Logs application results."""

    def __init__(self):
        self.session_id = f"session-{datetime.now().strftime('%Y%m%d-%H%M%S')}"
        self.results = []

    def log_result(self, job: dict, status: str, reason: str):
        result = {
            "jobId": job.get("jobId", ""),
            "company": job.get("company", ""),
            "jobTitle": job.get("title", ""),
            "linkedinUrl": job.get("linkedinUrl", ""),
            "applicationTime": datetime.now(timezone.utc).isoformat(),
            "status": status,
            "reason": reason,
            "session": self.session_id
        }
        self.results.append(result)
        self._save_to_applied()

    def _save_to_applied(self):
        data = {"version": "1.0", "applications": [], "sessions": []}
        if APPLIED_FILE.exists():
            data = json.loads(APPLIED_FILE.read_text())

        data["applications"].extend(self.results[-1:])
        APPLIED_FILE.write_text(json.dumps(data, indent=2, ensure_ascii=False))

    def print_summary(self):
        applied = sum(1 for r in self.results if r["status"] == "applied")
        review = sum(1 for r in self.results if r["status"] == "needs-human-review")
        skipped = sum(1 for r in self.results if r["status"] == "skipped")

        print(f"\n{'='*50}")
        print(f"Session Complete: {self.session_id}")
        print(f"{'='*50}")
        print(f"  Applied:      {applied}")
        print(f"  Needs Review: {review}")
        print(f"  Skipped:      {skipped}")
        print(f"  Total:        {len(self.results)}")


class LinkedInApplier:
    """Handles LinkedIn Easy Apply automation."""

    def __init__(self, page: Page, knowledge: KnowledgeBase, persona: PersonaInfo, unknown: UnknownQuestions):
        self.page = page
        self.knowledge = knowledge
        self.persona = persona
        self.unknown = unknown
        self.has_unknown_fields = False

    def apply_to_job(self, job: dict) -> tuple[str, str]:
        """Apply to a single job. Returns (status, reason)."""
        url = job.get("linkedinUrl", "")
        if not url:
            return "skipped", "No LinkedIn URL"

        print(f"\n[APPLYING] {job.get('company', 'Unknown')} - {job.get('title', 'Unknown')}")
        print(f"  URL: {url}")

        try:
            # Navigate to job page
            self.page.goto(url, timeout=30000)
            time.sleep(2)

            # Check for Easy Apply button
            easy_apply_btn = self.page.locator("button:has-text('Easy Apply')").first
            if not easy_apply_btn.is_visible(timeout=5000):
                return "needs-human-review", "No Easy Apply button found"

            # Check if already applied
            if self.page.locator("text='Applied'").first.is_visible(timeout=1000):
                return "skipped", "Already applied"

            # Click Easy Apply
            easy_apply_btn.click()
            time.sleep(2)

            # Process form steps
            return self._process_form()

        except PlaywrightTimeout:
            return "needs-human-review", "Page load timeout"
        except Exception as e:
            return "error", str(e)

    def _process_form(self) -> tuple[str, str]:
        """Process multi-step Easy Apply form."""
        max_steps = 10
        self.has_unknown_fields = False

        for step in range(max_steps):
            print(f"  [Step {step + 1}]")
            time.sleep(1)

            # Fill visible form fields
            self._fill_current_fields()

            # If we encountered unknown fields, stop and mark for review
            if self.has_unknown_fields:
                self._close_modal()
                return "needs-human-review", "Unknown questions - add to knowledge.json"

            # Check for submit button
            submit_btn = self.page.locator("button:has-text('Submit application')").first
            if submit_btn.is_visible(timeout=1000):
                submit_btn.click()
                time.sleep(2)

                # Verify success
                if self.page.locator("text='Application sent'").first.is_visible(timeout=5000):
                    self._close_modal()
                    return "applied", "Success"
                else:
                    return "needs-human-review", "Submit clicked but no confirmation"

            # Check for review button
            review_btn = self.page.locator("button:has-text('Review')").first
            if review_btn.is_visible(timeout=1000):
                review_btn.click()
                time.sleep(1)
                continue

            # Check for next button
            next_btn = self.page.locator("button:has-text('Next')").first
            if next_btn.is_visible(timeout=1000):
                next_btn.click()
                time.sleep(1)
                continue

            # No navigation button found
            return "needs-human-review", f"Stuck at step {step + 1}"

        return "needs-human-review", "Too many form steps"

    def _fill_current_fields(self):
        """Fill all visible form fields on current step."""
        # Text inputs
        for input_elem in self.page.locator("input[type='text'], input[type='tel'], input[type='email']").all():
            if input_elem.is_visible() and not input_elem.input_value():
                label = self._get_field_label(input_elem)
                if label:
                    answer = self._get_answer_for_field(label)
                    if answer:
                        input_elem.fill(answer)
                        print(f"    Filled '{label}': {answer}")

        # Dropdowns
        for select in self.page.locator("select").all():
            if select.is_visible():
                label = self._get_field_label(select)
                if label:
                    options = [opt.text_content() for opt in select.locator("option").all()]
                    answer = self._get_answer_for_field(label, options)
                    if answer:
                        try:
                            select.select_option(label=answer)
                            print(f"    Selected '{label}': {answer}")
                        except:
                            pass

        # Radio buttons - select first or infer
        radio_groups = {}
        for radio in self.page.locator("input[type='radio']").all():
            if radio.is_visible():
                name = radio.get_attribute("name")
                if name and name not in radio_groups:
                    radio_groups[name] = radio
                    label = self._get_field_label(radio)
                    # For now, click the radio to select it
                    # TODO: Better logic for choosing options
                    radio.click()
                    print(f"    Selected radio: {label or name}")

    def _get_field_label(self, elem) -> str:
        """Get label text for a form element."""
        # Try aria-label
        label = elem.get_attribute("aria-label")
        if label:
            return label

        # Try associated label
        elem_id = elem.get_attribute("id")
        if elem_id:
            label_elem = self.page.locator(f"label[for='{elem_id}']").first
            if label_elem.is_visible():
                return label_elem.text_content().strip()

        # Try parent label
        parent = elem.locator("xpath=ancestor::label").first
        if parent.count() > 0:
            return parent.text_content().strip()

        # Try placeholder
        placeholder = elem.get_attribute("placeholder")
        if placeholder:
            return placeholder

        return ""

    def _get_answer_for_field(self, question: str, options: list[str] = None) -> Optional[str]:
        """Get answer from knowledge or persona. Returns None if unknown."""
        # 1. Check knowledge base
        answer = self.knowledge.get_answer(question)
        if answer:
            return answer

        # 2. Check persona for common fields
        q_lower = question.lower()
        if "email" in q_lower:
            return self.persona.get_info("email")
        if "phone" in q_lower and "country" not in q_lower:
            return self.persona.get_info("phone")
        if "name" in q_lower and "first" not in q_lower and "last" not in q_lower:
            return self.persona.get_info("name")

        # 3. Unknown question - track it and mark job for review
        self.unknown.add(question, options)
        self.has_unknown_fields = True
        return None

    def _close_modal(self):
        """Close the Easy Apply modal."""
        self.page.keyboard.press("Escape")
        time.sleep(0.5)


def load_pending_jobs() -> list[dict]:
    """Load jobs from pending-jobs.json."""
    if not PENDING_FILE.exists():
        return []
    data = json.loads(PENDING_FILE.read_text())
    return data.get("jobs", [])


def remove_from_pending(job_id: str):
    """Remove a job from pending list."""
    if not PENDING_FILE.exists():
        return
    data = json.loads(PENDING_FILE.read_text())
    data["jobs"] = [j for j in data.get("jobs", []) if j.get("jobId") != job_id]
    data["lastUpdated"] = datetime.now(timezone.utc).isoformat()
    PENDING_FILE.write_text(json.dumps(data, indent=2, ensure_ascii=False))


def main():
    parser = argparse.ArgumentParser(description="LinkedIn Easy Apply Automation")
    parser.add_argument("--yolo", action="store_true", help="Apply without confirmation")
    parser.add_argument("--limit", type=int, default=0, help="Max jobs to apply")
    args = parser.parse_args()

    print("="*50)
    print("LinkedIn Easy Apply Automation")
    print("="*50)

    # Load resources
    try:
        knowledge = KnowledgeBase()
        persona = PersonaInfo()
        unknown = UnknownQuestions()
        logger = ApplicationLogger()
    except FileNotFoundError as e:
        print(f"[ERROR] {e}")
        print("Please run /moltoffer-candidate kickoff first.")
        sys.exit(1)

    # Load pending jobs
    jobs = load_pending_jobs()
    if not jobs:
        print("\n[INFO] No pending jobs. Run /moltoffer-candidate daily-match first.")
        sys.exit(0)

    if args.limit > 0:
        jobs = jobs[:args.limit]

    print(f"\nFound {len(jobs)} pending jobs.")
    print(f"Mode: {'YOLO (no confirmation)' if args.yolo else 'Interactive'}")

    # Start browser
    with sync_playwright() as p:
        print("\n[INFO] Launching browser...")
        print("[INFO] Please make sure you are logged into LinkedIn!")

        browser = p.chromium.launch(headless=False)
        context = browser.new_context()
        page = context.new_page()

        # Navigate to LinkedIn to check login
        page.goto("https://www.linkedin.com/feed/")
        time.sleep(3)

        if "login" in page.url or "signin" in page.url:
            print("\n[WARN] Not logged into LinkedIn!")
            print("Please log in manually, then press Enter to continue...")
            input()

        applier = LinkedInApplier(page, knowledge, persona, unknown)

        # Process jobs
        for i, job in enumerate(jobs):
            print(f"\n[{i+1}/{len(jobs)}] Processing...")

            if not args.yolo:
                print(f"  Company: {job.get('company', 'Unknown')}")
                print(f"  Title: {job.get('title', 'Unknown')}")
                response = input("  Apply? [y/n/q]: ").strip().lower()
                if response == "q":
                    break
                if response != "y":
                    logger.log_result(job, "skipped", "User skipped")
                    continue

            status, reason = applier.apply_to_job(job)
            logger.log_result(job, status, reason)

            if status == "applied":
                remove_from_pending(job.get("jobId", ""))

            print(f"  Result: {status} - {reason}")

            # Rate limiting
            if i < len(jobs) - 1:
                delay = 3 if args.yolo else 1
                time.sleep(delay)

        browser.close()

    logger.print_summary()
    unknown.print_summary()


if __name__ == "__main__":
    main()

File: skills/moltoffer-auto-apply/scripts/auto_apply.js
#!/usr/bin/env node
/**
 * LinkedIn Easy Apply Automation Script
 *
 * This script automates LinkedIn Easy Apply using Playwright.
 * AI is only called when encountering new questions not in knowledge.json.
 *
 * Usage:
 *   node auto_apply.js [--yolo] [--limit N]
 *
 * Options:
 *   --yolo      Apply without confirmation for each job
 *   --limit N   Maximum number of jobs to apply (default: unlimited)
 */

import { chromium } from 'playwright';
import Anthropic from '@anthropic-ai/sdk';
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import * as readline from 'readline';
import YAML from 'yaml';

// Paths
const __filename = fileURLToPath(import.meta.url);
const SCRIPT_DIR = dirname(__filename);
const SKILL_DIR = join(SCRIPT_DIR, '..');
const DATA_DIR = join(SKILL_DIR, 'data');
const CANDIDATE_DIR = join(SKILL_DIR, '..', 'moltoffer-candidate');

// Files
const KNOWLEDGE_FILE = join(DATA_DIR, 'knowledge.json');
const APPLIED_FILE = join(DATA_DIR, 'applied.json');
const PENDING_FILE = join(DATA_DIR, 'pending-jobs.json');
const LOGS_FILE = join(DATA_DIR, 'logs.json');
const PERSONA_FILE = join(CANDIDATE_DIR, 'persona.md');
const CREDENTIALS_FILE = join(CANDIDATE_DIR, 'credentials.local.json');

// Helper to read JSON file
function readJSON(filepath, defaultValue = {}) {
  if (!existsSync(filepath)) return defaultValue;
  return JSON.parse(readFileSync(filepath, 'utf-8'));
}

// Helper to write JSON file
function writeJSON(filepath, data) {
  writeFileSync(filepath, JSON.stringify(data, null, 2), 'utf-8');
}

// Helper for user input
function prompt(question) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer.trim().toLowerCase());
    });
  });
}

/**
 * Manages form answers and learns from AI responses.
 */
class KnowledgeBase {
  constructor() {
    this.data = this._load();
  }

  _load() {
    return readJSON(KNOWLEDGE_FILE, {
      version: '1.0',
      commonAnswers: {},
      assumptions: []
    });
  }

  save() {
    writeJSON(KNOWLEDGE_FILE, this.data);
  }

  getAnswer(question) {
    const qLower = question.toLowerCase();

    // Check common answers
    for (const [key, value] of Object.entries(this.data.commonAnswers || {})) {
      if (key.toLowerCase().includes(qLower) || qLower.includes(key.toLowerCase())) {
        return value;
      }
    }

    // Check assumptions
    for (const assumption of this.data.assumptions || []) {
      if ((assumption.question || '').toLowerCase() === qLower) {
        return assumption.assumedAnswer;
      }
    }

    return null;
  }

  addAssumption(question, answer, reasoning) {
    if (!this.data.assumptions) this.data.assumptions = [];
    this.data.assumptions.push({
      question,
      assumedAnswer: answer,
      reasoning,
      timestamp: new Date().toISOString()
    });
    this.save();
    console.log(`  [LEARNED] Q: ${question.substring(0, 50)}... A: ${answer}`);
  }
}

/**
 * Extracts info from persona.md for form filling.
 */
class PersonaInfo {
  constructor() {
    this.raw = '';
    this.frontmatter = {};
    this._load();
  }

  _load() {
    if (!existsSync(PERSONA_FILE)) {
      throw new Error(`Persona file not found: ${PERSONA_FILE}`);
    }

    this.raw = readFileSync(PERSONA_FILE, 'utf-8');

    // Parse YAML frontmatter
    if (this.raw.startsWith('---')) {
      const parts = this.raw.split('---');
      if (parts.length >= 3) {
        this.frontmatter = YAML.parse(parts[1]) || {};
      }
    }
  }

  getInfo(key) {
    // Check frontmatter first
    if (key in this.frontmatter) {
      return String(this.frontmatter[key]);
    }

    // Search in raw content
    const patterns = {
      name: /###\s*Name\s*\n+([^\n]+)/,
      email: /[\w.+-]+@[\w-]+\.[\w.-]+/,
      phone: /\+?[\d\s()-]{10,}/,
      location: /###\s*Current Location\s*\n+([^\n]+)/
    };

    if (key in patterns) {
      const match = this.raw.match(patterns[key]);
      if (match) {
        return match[1] || match[0];
      }
    }

    return null;
  }
}

/**
 * Calls Claude API only when needed for new questions.
 */
class AIAssistant {
  constructor(persona) {
    this.client = null;
    this.persona = persona;
    this._initClient();
  }

  _initClient() {
    const apiKey = process.env.ANTHROPIC_API_KEY;
    if (!apiKey) {
      console.log('[WARN] ANTHROPIC_API_KEY not set. AI assistance disabled.');
      return;
    }
    this.client = new Anthropic({ apiKey });
  }

  async inferAnswer(question, options = null) {
    if (!this.client) {
      return [this._guessDefault(question, options), 'default (AI unavailable)'];
    }

    const prompt = `Based on this candidate profile, answer the following job application question.

CANDIDATE PROFILE:
${this.persona.raw.substring(0, 2000)}

QUESTION: ${question}
${options ? `OPTIONS: ${options.join(', ')}` : ''}

Respond with ONLY the answer value (no explanation). If it's a yes/no question, respond with just "Yes" or "No".
If there are options, respond with the exact option text.
If it's a number, respond with just the number.`;

    try {
      const response = await this.client.messages.create({
        model: 'claude-sonnet-4-20250514',
        max_tokens: 100,
        messages: [{ role: 'user', content: prompt }]
      });
      const answer = response.content[0].text.trim();
      return [answer, 'AI inference'];
    } catch (e) {
      console.log(`  [WARN] AI error: ${e.message}`);
      return [this._guessDefault(question, options), 'default (AI error)'];
    }
  }

  _guessDefault(question, options = null) {
    const qLower = question.toLowerCase();

    // Yes/No defaults
    if (['willing', 'able', 'authorized', 'agree'].some(x => qLower.includes(x))) {
      return 'Yes';
    }
    if (['convicted', 'disability', 'veteran'].some(x => qLower.includes(x))) {
      return 'No';
    }

    // If options provided, pick first one
    if (options && options.length > 0) {
      return options[0];
    }

    // Number defaults
    if (qLower.includes('years') || qLower.includes('experience')) {
      return '3';
    }

    return '';
  }
}

/**
 * Logs application results.
 */
class ApplicationLogger {
  constructor() {
    this.sessionId = `session-${new Date().toISOString().replace(/[:.]/g, '-').substring(0, 19)}`;
    this.results = [];
  }

  logResult(job, status, reason) {
    const result = {
      jobId: job.jobId || '',
      company: job.company || '',
      jobTitle: job.title || '',
      linkedinUrl: job.linkedinUrl || '',
      applicationTime: new Date().toISOString(),
      status,
      reason,
      session: this.sessionId
    };
    this.results.push(result);
    this._saveToApplied();
  }

  _saveToApplied() {
    let data = readJSON(APPLIED_FILE, { version: '1.0', applications: [], sessions: [] });
    data.applications.push(this.results[this.results.length - 1]);
    writeJSON(APPLIED_FILE, data);
  }

  printSummary() {
    const applied = this.results.filter(r => r.status === 'applied').length;
    const review = this.results.filter(r => r.status === 'needs-human-review').length;
    const skipped = this.results.filter(r => r.status === 'skipped').length;

    console.log(`\n${'='.repeat(50)}`);
    console.log(`Session Complete: ${this.sessionId}`);
    console.log('='.repeat(50));
    console.log(`  Applied:      ${applied}`);
    console.log(`  Needs Review: ${review}`);
    console.log(`  Skipped:      ${skipped}`);
    console.log(`  Total:        ${this.results.length}`);
  }
}

/**
 * Handles LinkedIn Easy Apply automation.
 */
class LinkedInApplier {
  constructor(page, knowledge, persona, ai) {
    this.page = page;
    this.knowledge = knowledge;
    this.persona = persona;
    this.ai = ai;
  }

  async applyToJob(job) {
    const url = job.linkedinUrl || '';
    if (!url) {
      return ['skipped', 'No LinkedIn URL'];
    }

    console.log(`\n[APPLYING] ${job.company || 'Unknown'} - ${job.title || 'Unknown'}`);
    console.log(`  URL: ${url}`);

    try {
      // Navigate to job page
      await this.page.goto(url, { timeout: 30000 });
      await this.page.waitForTimeout(2000);

      // Check for Easy Apply button
      const easyApplyBtn = this.page.locator("button:has-text('Easy Apply')").first();
      if (!await easyApplyBtn.isVisible({ timeout: 5000 }).catch(() => false)) {
        return ['needs-human-review', 'No Easy Apply button found'];
      }

      // Check if already applied
      const alreadyApplied = this.page.locator("text='Applied'").first();
      if (await alreadyApplied.isVisible({ timeout: 1000 }).catch(() => false)) {
        return ['skipped', 'Already applied'];
      }

      // Click Easy Apply
      await easyApplyBtn.click();
      await this.page.waitForTimeout(2000);

      // Process form steps
      return await this._processForm();
    } catch (e) {
      if (e.name === 'TimeoutError') {
        return ['needs-human-review', 'Page load timeout'];
      }
      return ['error', e.message];
    }
  }

  async _processForm() {
    const maxSteps = 10;

    for (let step = 0; step < maxSteps; step++) {
      console.log(`  [Step ${step + 1}]`);
      await this.page.waitForTimeout(1000);

      // Fill visible form fields
      await this._fillCurrentFields();

      // Check for submit button
      const submitBtn = this.page.locator("button:has-text('Submit application')").first();
      if (await submitBtn.isVisible({ timeout: 1000 }).catch(() => false)) {
        await submitBtn.click();
        await this.page.waitForTimeout(2000);

        // Verify success
        const success = this.page.locator("text='Application sent'").first();
        if (await success.isVisible({ timeout: 5000 }).catch(() => false)) {
          await this._closeModal();
          return ['applied', 'Success'];
        } else {
          return ['needs-human-review', 'Submit clicked but no confirmation'];
        }
      }

      // Check for review button
      const reviewBtn = this.page.locator("button:has-text('Review')").first();
      if (await reviewBtn.isVisible({ timeout: 1000 }).catch(() => false)) {
        await reviewBtn.click();
        await this.page.waitForTimeout(1000);
        continue;
      }

      // Check for next button
      const nextBtn = this.page.locator("button:has-text('Next')").first();
      if (await nextBtn.isVisible({ timeout: 1000 }).catch(() => false)) {
        await nextBtn.click();
        await this.page.waitForTimeout(1000);
        continue;
      }

      // No navigation button found
      return ['needs-human-review', `Stuck at step ${step + 1}`];
    }

    return ['needs-human-review', 'Too many form steps'];
  }

  async _fillCurrentFields() {
    // Text inputs
    const inputs = await this.page.locator("input[type='text'], input[type='tel'], input[type='email']").all();
    for (const input of inputs) {
      if (await input.isVisible() && !(await input.inputValue())) {
        const label = await this._getFieldLabel(input);
        if (label) {
          const answer = await this._getAnswerForField(label);
          if (answer) {
            await input.fill(answer);
            console.log(`    Filled '${label}': ${answer}`);
          }
        }
      }
    }

    // Dropdowns
    const selects = await this.page.locator('select').all();
    for (const select of selects) {
      if (await select.isVisible()) {
        const label = await this._getFieldLabel(select);
        if (label) {
          const optionElements = await select.locator('option').all();
          const options = await Promise.all(optionElements.map(opt => opt.textContent()));
          const answer = await this._getAnswerForField(label, options);
          if (answer) {
            try {
              await select.selectOption({ label: answer });
              console.log(`    Selected '${label}': ${answer}`);
            } catch (e) {
              // Ignore selection errors
            }
          }
        }
      }
    }

    // Radio buttons - select first in each group
    const radios = await this.page.locator("input[type='radio']").all();
    const radioGroups = new Set();
    for (const radio of radios) {
      if (await radio.isVisible()) {
        const name = await radio.getAttribute('name');
        if (name && !radioGroups.has(name)) {
          radioGroups.add(name);
          const label = await this._getFieldLabel(radio);
          await radio.click();
          console.log(`    Selected radio: ${label || name}`);
        }
      }
    }
  }

  async _getFieldLabel(elem) {
    // Try aria-label
    let label = await elem.getAttribute('aria-label');
    if (label) return label;

    // Try associated label
    const elemId = await elem.getAttribute('id');
    if (elemId) {
      const labelElem = this.page.locator(`label[for='${elemId}']`).first();
      if (await labelElem.isVisible().catch(() => false)) {
        const text = await labelElem.textContent();
        if (text) return text.trim();
      }
    }

    // Try placeholder
    const placeholder = await elem.getAttribute('placeholder');
    if (placeholder) return placeholder;

    return '';
  }

  async _getAnswerForField(question, options = null) {
    // 1. Check knowledge base
    let answer = this.knowledge.getAnswer(question);
    if (answer) return answer;

    // 2. Check persona for common fields
    const qLower = question.toLowerCase();
    if (qLower.includes('email')) {
      return this.persona.getInfo('email') || '';
    }
    if (qLower.includes('phone') && !qLower.includes('country')) {
      return this.persona.getInfo('phone') || '';
    }
    if (qLower.includes('name') && !qLower.includes('first') && !qLower.includes('last')) {
      return this.persona.getInfo('name') || '';
    }

    // 3. Ask AI and learn
    const [inferredAnswer, reasoning] = await this.ai.inferAnswer(question, options);
    if (inferredAnswer) {
      this.knowledge.addAssumption(question, inferredAnswer, reasoning);
    }

    return inferredAnswer;
  }

  async _closeModal() {
    await this.page.keyboard.press('Escape');
    await this.page.waitForTimeout(500);
  }
}

/**
 * Load jobs from pending-jobs.json
 */
function loadPendingJobs() {
  const data = readJSON(PENDING_FILE, { jobs: [] });
  return data.jobs || [];
}

/**
 * Remove a job from pending list
 */
function removeFromPending(jobId) {
  const data = readJSON(PENDING_FILE, { jobs: [] });
  data.jobs = (data.jobs || []).filter(j => j.jobId !== jobId);
  data.lastUpdated = new Date().toISOString();
  writeJSON(PENDING_FILE, data);
}

/**
 * Parse command line arguments
 */
function parseArgs() {
  const args = process.argv.slice(2);
  const result = { yolo: false, limit: 0 };

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--yolo') {
      result.yolo = true;
    } else if (args[i] === '--limit' && args[i + 1]) {
      result.limit = parseInt(args[i + 1], 10);
      i++;
    }
  }

  return result;
}

/**
 * Main function
 */
async function main() {
  const args = parseArgs();

  console.log('='.repeat(50));
  console.log('LinkedIn Easy Apply Automation');
  console.log('='.repeat(50));

  // Load resources
  let knowledge, persona, ai, logger;
  try {
    knowledge = new KnowledgeBase();
    persona = new PersonaInfo();
    ai = new AIAssistant(persona);
    logger = new ApplicationLogger();
  } catch (e) {
    console.log(`[ERROR] ${e.message}`);
    console.log('Please run /moltoffer-candidate kickoff first.');
    process.exit(1);
  }

  // Load pending jobs
  let jobs = loadPendingJobs();
  if (jobs.length === 0) {
    console.log('\n[INFO] No pending jobs. Run /moltoffer-candidate daily-match first.');
    process.exit(0);
  }

  if (args.limit > 0) {
    jobs = jobs.slice(0, args.limit);
  }

  console.log(`\nFound ${jobs.length} pending jobs.`);
  console.log(`Mode: ${args.yolo ? 'YOLO (no confirmation)' : 'Interactive'}`);

  // Start browser
  console.log('\n[INFO] Launching browser...');
  console.log('[INFO] Please make sure you are logged into LinkedIn!');

  const browser = await chromium.launch({ headless: false });
  const context = await browser.newContext();
  const page = await context.newPage();

  // Navigate to LinkedIn to check login
  await page.goto('https://www.linkedin.com/feed/');
  await page.waitForTimeout(3000);

  if (page.url().includes('login') || page.url().includes('signin')) {
    console.log('\n[WARN] Not logged into LinkedIn!');
    console.log('Please log in manually, then press Enter to continue...');
    await prompt('');
  }

  const applier = new LinkedInApplier(page, knowledge, persona, ai);

  // Process jobs
  for (let i = 0; i < jobs.length; i++) {
    const job = jobs[i];
    console.log(`\n[${i + 1}/${jobs.length}] Processing...`);

    if (!args.yolo) {
      console.log(`  Company: ${job.company || 'Unknown'}`);
      console.log(`  Title: ${job.title || 'Unknown'}`);
      const response = await prompt('  Apply? [y/n/q]: ');
      if (response === 'q') {
        break;
      }
      if (response !== 'y') {
        logger.logResult(job, 'skipped', 'User skipped');
        continue;
      }
    }

    const [status, reason] = await applier.applyToJob(job);
    logger.logResult(job, status, reason);

    if (status === 'applied') {
      removeFromPending(job.jobId || '');
    }

    console.log(`  Result: ${status} - ${reason}`);

    // Rate limiting
    if (i < jobs.length - 1) {
      const delay = args.yolo ? 3000 : 1000;
      await page.waitForTimeout(delay);
    }
  }

  await browser.close();
  logger.printSummary();
}

main().catch(console.error);
